x/orderbook/model.go

```diff
@@ -0,0 +1,270 @@
+		ID:       copyBytes(m.ID),
+		Owner:    m.Owner.Clone(),
+		Name:     m.Name,
+	}
+}
+
+// Validate is always succesful
+func (m *Market) Validate() error {
+	if err := m.Metadata.Validate(); err != nil {
+		return err
+	}
+	if err := isGenID(m.ID, true); err != nil {
+		return err
+	}
+	if err := m.Owner.Validate(); err != nil {
+		return errors.Wrap(err, "owner")
+	}
+	if !validMarketName(m.Name) {
+		return errors.Wrap(errors.ErrModel, "invalid market name")
+	}
+	return nil
+}
+
+var _ morm.Model = (*OrderBook)(nil)
+
+// SetID is a minimal implementation, useful when the ID is a separate protobuf field
+func (o *OrderBook) SetID(id []byte) error {
+	o.ID = id
+	return nil
+}
+
+// Copy produces a new copy to fulfill the Model interface
+func (o *OrderBook) Copy() orm.CloneableData {
+	return &OrderBook{
+		Metadata:      o.Metadata.Copy(),
+		ID:            copyBytes(o.ID),
+		MarketID:      copyBytes(o.MarketID),
+		AskTicker:     o.AskTicker,
+		BidTicker:     o.BidTicker,
+		TotalAskCount: o.TotalAskCount,
+		TotalBidCount: o.TotalBidCount,
+	}
+}
+
+// Validate is always succesful
+func (o *OrderBook) Validate() error {
+	if err := o.Metadata.Validate(); err != nil {
+		return err
+	}
+	if err := isGenID(o.ID, true); err != nil {
+		return err
+	}
+	if err := isGenID(o.MarketID, false); err != nil {
+		return errors.Wrap(err, "market id")
+	}
+	if !coin.IsCC(o.AskTicker) {
+		return errors.Wrap(errors.ErrModel, "invalid ask ticker")
+	}
+	if !coin.IsCC(o.BidTicker) {
+		return errors.Wrap(errors.ErrModel, "invalid bid ticker")
+	}
+	if o.TotalAskCount < 0 {
+		return errors.Wrap(errors.ErrModel, "negative total ask count")
+	}
+	if o.TotalBidCount < 0 {
+		return errors.Wrap(errors.ErrModel, "negative total bid count")
+	}
+	return nil
+}
+
+var _ morm.Model = (*Order)(nil)
+
+// SetID is a minimal implementation, useful when the ID is a separate protobuf field
+func (o *Order) SetID(id []byte) error {
+	o.ID = id
+	return nil
+}
+
+// Copy produces a new copy to fulfill the Model interface
+func (o *Order) Copy() orm.CloneableData {
+	return &Order{
+		Metadata:       o.Metadata.Copy(),
+		ID:             copyBytes(o.ID),
+		Trader:         copyBytes(o.Trader),
+		OrderBookID:    copyBytes(o.OrderBookID),
+		Side:           o.Side,
+		OrderState:     o.OrderState,
+		OriginalOffer:  o.OriginalOffer.Clone(),
+		RemainingOffer: o.RemainingOffer.Clone(),
+		Price:          o.Price.Clone(),
+		TradeIds:       copyBytesList(o.TradeIds),
+		CreatedAt:      o.CreatedAt,
+		UpdatedAt:      o.UpdatedAt,
+	}
+}
+
+// Validate is always succesful
+func (o *Order) Validate() error {
+	if err := o.Metadata.Validate(); err != nil {
+		return err
+	}
+	if err := isGenID(o.ID, true); err != nil {
+		return err
+	}
+	if err := o.Trader.Validate(); err != nil {
+		return errors.Wrap(err, "trader")
+	}
+	if err := isGenID(o.OrderBookID, false); err != nil {
+		return errors.Wrap(err, "order book id")
+	}
+	// TODO: valid Side?
```

Comment:
```SIDE_INVALID = 0 [(gogoproto.enumvalue_customname) = "Invalid"];``` [Side](https://github.com/iov-one/tutorial/blob/master/x/orderbook/codec.proto#L39) has an invalid enum value which is not necessary but forced by ```protolint```. Maybe we can add a validation for ensuring side is not invalid?

Reply:
Yup. We should.
I was wondering how to enforce that it was either ASK or BID and not just some other number (17). 

The definition is `type Side int32`, so the compiler will not complain about

```go
Order{
   // ...
  Side: int32(17),
  // ...
}
```

Only if we mixed up eg. Side and OrderState